ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs

```diff
@@ -305,30 +308,73 @@ applyBlock :: forall m c l blk
             . (ApplyBlock l blk, TickedTableStuff l, Monad m, c)
            => LedgerCfg l
            -> Ap m l blk c
-           -> LedgerDB l -> m (l TrackingMK)
+           -> LedgerDB l
+           -> m (l ValuesMK, l TrackingMK)
 applyBlock cfg ap db = case ap of
-    ReapplyVal b ->
+    Weaken ap' ->
+        applyBlock cfg ap' db
+    _ -> do
+      old <- oldApplyBlock ap
+      new <- newApplyBlock ap
+
+      assert (old == applyTracking oldLedgerDbCurrent new) $ return (old, new)
+
+  where
+
+    oldLedgerDbCurrent :: l ValuesMK
+    oldLedgerDbCurrent = either unCheckpoint unCheckpoint . AS.head . ledgerDbCheckpoints $ db
+
+    oldApplyBlock :: Ap m l blk c -> m (l ValuesMK)
+    oldApplyBlock = \case
+      ReapplyVal b -> return $ forgetLedgerStateTracking $ tickThenReapply cfg b oldLedgerDbCurrent
+
+      ApplyVal b -> forgetLedgerStateTracking <$>
+               (  either (throwLedgerError db (blockRealPoint b)) return
+                $ runExcept
+                $ tickThenApply cfg b oldLedgerDbCurrent)
+
+      ReapplyRef r  -> do
+        b <- resolveBlock r -- TODO: ask: would it make sense to recursively call applyBlock using ReapplyVal?
+
+        return $ forgetLedgerStateTracking $ tickThenReapply cfg b oldLedgerDbCurrent
+
+      ApplyRef r -> do
+        b <- resolveBlock r
+
+        forgetLedgerStateTracking <$>
+             ( either (throwLedgerError db (blockRealPoint b)) return
+             $ runExcept
+             $ tickThenApply cfg b oldLedgerDbCurrent)
+
+      Weaken _ -> error "unreachable"
```

Comment:
Maybe we could say that we don't expect `ap` to contain nested `Weaken`, however, I worry that we're depending on an implementation detail from somewhere else. Why not simply use the old pattern in which `Weaken` would cause the function to recurse?